consensus/src/messages/handlers.rs

```diff
@@ -304,144 +304,190 @@ impl<N: Network> Handle<N, ResponseChunk, Arc<RwLock<Blockchain>>> for RequestCh
+        log::info!("Could not find the desired block to create the txn proof");
+        return ResponseTransactionsProof {
+            proof: None,
+            block: None,
+        };
+    }
 
-            // If we were provided a block number corresponding to a checkpoint block, it needs to correspond to the current epoch
-            // Otherwise, the requester should use the latest epoch number.
-            if self.block_number < election_head {
-                log::info!(
-                    block_number = self.block_number,
-                    "Requested txn proof that corresponds to a finalized epoch, should use the election block instead",
-                );
-                return ResponseTransactionsProof {
-                    proof: None,
-                    block: None,
-                };
-            }
+    let proof = blockchain.history_store.prove(
+        Policy::epoch_at(block_number),
+        hashes,
+        verifier_state,
+        None,
+    );
+
+    // If we couldn't obtain a proof we return None
+    if proof.is_none() {
+        log::info!("Could not generate the txn inclusion proof");
+        return ResponseTransactionsProof {
+            proof: None,
+            block: None,
+        };
+    }
 
-            block = blockchain
-                .chain_store
-                .get_block_at(self.block_number, false, None)
-                .ok();
+    ResponseTransactionsProof { proof, block }
+}
+
+#[cfg(feature = "full")]
+fn prove_transaction(
+    blockchain: &Arc<RwLock<Blockchain>>,
+    transaction: &Blake2bHash,
+) -> ResponseTransactionsProof {
+    let blockchain = blockchain.read();
+    //let hashes: Vec<&Blake2bHash> = transactions.iter().collect();
+
+    let mut verifier_state = None;
+    let election_head = blockchain.election_head().block_number();
+    let macro_head = blockchain.macro_head().block_number();
+    let current_head = blockchain.head().block_number();
+
+    // Get the extended transaction from the history store
+    let mut extended_transactions = blockchain
+        .history_store
+        .get_ext_tx_by_hash(transaction, None);
+
+    // Due to the history store implementation, potentially, we could have multiple extended transactions at this hash
+    // So we just pick any transaction
+    if let Some(ext_txn) = extended_transactions.pop() {
+        let block_number = ext_txn.block_number;
+
+        let proving_block_number = if block_number <= election_head {
+            // If the txn is in a finalized epoch, we use the last election block
+            election_head
+        } else if block_number <= macro_head {
+            // If the txn is in a finalized batch in the current epoch, we use the last checkpoint block
+            macro_head
+        } else {
+            // If the txn is in the current batch, we use the head.
+            // This might be problematic if the requester is behind and does not have the current head (yet)
+            current_head
```

Comment:
Instead of the current head, could use the `ext_tx.block_number`. If it's in the current batch, the requester should have it.